cpp/src/arrow/util/bit_util_extract.cc

```diff
@@ -0,0 +1,247 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include <cstdint>
+
+#include "arrow/util/bit_util_extract.h"
+
+namespace arrow::bit_util {
+
+// clang-format off
+/* Python code to generate lookup table:
```

Comment:
`kPextTable` as a raw C array is entirely contiguous. Is it also the case for a nested `std::array`?

Also, what is the compilation time of your suggestion?

Reply:
std::array is a *very* thin wrapper around raw C arrays, so thin that it will not initialize trivial elements unless instructed:
```c++
int c[10]; // no initializer, array elements not zeroed
std::array<int, 10> cxx; // also
```

Compilation is fast enough that I can't get a consistent number for this example by itself; I'd need to rewrite it to a larger loop or something. Typically 8-16ms on my machine, which is the same range I see for the large array literal. In addition we only compile this once since we're not in a header

Reply:
The Python code exists here as documentation of where the table comes from. It's not something we plan to keep changing. If I replace the hardcoded constant table with this C++ code I will have to add tests to ensure the code is correct and these tests will have to include the table. And worse, I will have to expose this `constexpr` in a way that is accessible to tests.